pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerFactory.java

```diff
@@ -84,6 +82,8 @@ private SegmentPrunerFactory() {
         }
       }
     }
+    // Always prune out empty segments first
```

Comment:
Why move it to the bottom?

Reply:
For this change I am a bit concerned whether this change will break some of the existing features, that we may end up selecting some empty segments and later got pruned out, resulting a failed query .

The move is after the below finding:
In our tests, we found that `EmptySegmentPruner` is trying to create a duplicated copy of all the segment names every time. 
It shows up in the jstack outputs, meaning that copying 30k+ segment names for each query is actually not cheap. Moving it to the bottom may mean better performance, in our test it is the difference of **60k QPS => 80k QPS**
(in the case if there is empty segment around)

Reply:
> For the partition pruning, since in most cases a table will use the same partition function for all segments, we can group the segments in a different way:
> 
> * Map<PartitionFunction, Map<Integer, Set>>: partition function -> partition id -> segments
> 
> When getting a query, we can loop over the partition functions from this map and calculate all the matching segments. Then we loop over the input segments and keep the segments that does not have partition info, or included in the matching segments.

Good suggestion; the major challenge of this approach would be analyzing the `isPartitionMatch(Expression filterExpression, PartitionInfo partitionInfo)` method to pick out the segments properly. In a SQL filter we could have something like `WHERE account_key in ['ac23d', 'e92df'] OR account_key = 'c3f9d'`, without proper rewrite it will be hard to make our recursive selection right.

For now my code has become a targeted optimization mostly for Offline table queries, because the externalView is much more stable.